crates/sui-graphql-rpc/src/types/dot_move/named_type.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use async_graphql::Context;
+use futures::future;
+use move_core_types::parser::parse_type_tag;
+use sui_types::base_types::ObjectID;
+
+use crate::error::Error;
+
+use super::{
+    config::{DotMoveServiceError, VERSIONED_NAME_UNBOUND_REG},
+    named_move_package::NamedMovePackage,
+};
+
+pub(crate) struct NamedType;
+
+impl NamedType {
+    pub(crate) async fn query(
+        ctx: &Context<'_>,
+        name: &str,
+        checkpoint_viewed_at: u64,
+    ) -> Result<String, Error> {
+        // we do not de-duplicate the names here, as the dataloader will do this for us.
+        let names = Self::parse_names(name)?;
+
+        // Gather all the requests to resolve the names.
+        let names_to_resolve = names
+            .iter()
+            .map(|x| NamedMovePackage::query(ctx, x, checkpoint_viewed_at))
+            .collect::<Vec<_>>();
+
+        // now we resolve all the names in parallel (data-loader will do the proper de-duplication / batching for us)
+        // also the `NamedMovePackage` query will re-validate the names (including max length, which is not checked on the regex).
+        let mut results = future::try_join_all(names_to_resolve).await?;
+
+        // now let's create a hashmap with {name: MovePackage}
+        let mut name_package_id_mapping = HashMap::new();
```

Comment:
`DataLoader` offers a `load_many` API that will do roughly this -- it may be worthwhile to expose access to that via a function on `NamedMovePackage`. I did something similar for `TransactionBlock` here:

https://github.com/MystenLabs/sui/blob/1bd2379ccc6f769b79d444035fccf558d77a14d3/crates/sui-graphql-rpc/src/types/transaction_block.rs#L270-L289

Reply:
Discussed offline, gets kinda confusing since we're using other data loaders internally too.